File: my-alogorithms-study/src/main/scala/com/wenthomas/alogorithms/sort/QuickSort.scala
package com.wenthomas.alogorithms.sort

import scala.util.Random

/**
 * @author Verno
 * @create 2020-04-11 14:27 
 */
/**
 * 快速排序：针对数据量较大的场景效果比较明显
 */
object QuickSort {
    def main(args: Array[String]): Unit = {
        var arr1 = Array(5,4,3,2,1)
        sort(arr1)
        println(arr1.mkString(","))

/*        val arr = randomArr()
        sort(arr)
        println(arr.mkString(","))*/

        //val arr1 = scalaQuickSort(arr)
        //println(arr1.mkString(","))

        //println(scalaQuickSortList(arr.toList).take(100).mkString(","))


        //val arr2 = Array(5,4,3,2,1)
        val arr2 = randomArr()
        val array = scalaQuickSortWithPartition(arr2)
        println(array.mkString(","))


    }

    def randomArr() = {
        val random = new Random()
        (1 to 2000).map(_ => random.nextInt(1000)).toArray
    }

    def swap(arr: Array[Int], a: Int, b: Int) = {
        val tmp = arr(a)
        arr(a) = arr(b)
        arr(b) = tmp
    }
    /**
     * 快速排序：递归实现，原地排序
     * @param arr
     */
    def sort(arr: Array[Int]) = {

        //分区：分为左区和右区
        def partition(array: Array[Int], left: Int, right: Int) = {
            //左右指针
            var low = left
            var high = right
            //基准值
            val zero = arr(left)

            while (low < high) {
                //注意：low <= high 与 arr(low)的顺序不能调换，否则某些情况会导致下角标越界异常，例如（5,4,3,2,1）序列
                while (low <= right && arr(low) <= zero) {
                    low += 1
                }

                while (left <= high && arr(high) > zero) {
                    high -= 1
                }

                if (low < high) {
                    swap(arr, low, high)
                }
            }

            //返回基准值的正确位置
            swap(arr, left, high)

            high
        }

        //排序，left开始的坐标，right结束的坐标  [left, right]
        def quick(arr: Array[Int], left: Int, right: Int): Unit = {
            //当左指针到了右指针右边，结束
            if (left >= right) return

            //基准值的位置
            val mid = partition(arr, left, right)
            //对左边进行排序
            quick(arr, left, mid - 1)
            //对右边进行排序
            quick(arr, mid + 1, right)
        }

        quick(arr, 0, arr.length - 1)
    }

    /**
     * 快速排序：返回新数组
     * @param arr
     * @return
     */
    def scalaQuickSort(arr: Array[Int]): Array[Int] = {
        arr match {
            case Array(p, rest@_*) =>
                //找到小于p的所有元素
                val left = scalaQuickSort(rest.filter(_ <= p).toArray)
                //找到大于p的所有元素
                val right = scalaQuickSort(rest.filter(_ > p).toArray)

                (left :+ p) ++ right
            case _ => Array()
        }
    }


    def scalaQuickSortList(list: List[Int]): List[Int] = {
        list match {
            case p :: rest =>
                scalaQuickSortList(rest.filter(_ <= p)) ::: p ::scalaQuickSortList(rest.filter(_ > p))
            case Nil => Nil
        }
    }

    /**
     * 快速排序：使用partition算子进行左右分区
     * @param arr
     * @return
     */
    def scalaQuickSortWithPartition(arr: Array[Int]): Array[Int] = {
        arr match {
            case Array(p, rest@_*) =>
                val (left, right)= rest.partition(_ <= p)
                (scalaQuickSortWithPartition(left.toArray) :+ p) ++ scalaQuickSortWithPartition(right.toArray)
            case _ => Array()
        }
    }
}

File: my-alogorithms-study/src/main/java/com/wenthomas/alogorithms/QuickSort.java
package com.wenthomas.alogorithms;

import java.util.Arrays;

/**
 * 快速排序
 * tips:
 * 选定一个基准值pivot（通常指定为数组第一个值），比基准值大的放在右侧，比基准值小的放在左侧
 * 指定左右两个指针分别为left，right ；left < right
 * 指定函数退出条件，即left > right
 * 先从右向左遍历，即右指针向左移动——right–操作，发现比pivot小的值暂停移动
 * 再从左向右遍历，即左指针向右移动——left++操作，发现比pivot大的值暂停移动
 * 交换左右指针发现的两个值的位置
 * 当两指针相遇，即left == right，当前值与pivot交换位置
 *
 * 递归处理
 * ————————————————
 * @author Verno
 * @create 2022-11-28 14:48
 */
public class QuickSort {
    public static void quickSort(int[] array,int low, int high) {

        int right = high;
        int left = low;
        int pivot = array[low];
        int temp;

        while (left < right) {
            while (left < right && array[right] >= pivot) {
                right --;
            }
            while (left < right && array[left] <= pivot) {
                left ++;
            }
            temp = array[left];
            array[left] = array[right];
            array[right] = temp;

            if (left == right) {
                array[low] = array[left];
                array[left] = pivot;
            }
        }

        //递归左数组与右数组
        if (low < left) {
            quickSort(array, low, left - 1);
        }
        if (right < high) {
            quickSort(array, right + 1, high);
        }

    }


    public static void main(String[] args) {
        int[] array = {21, 34, 556, 12, 3324, 75, 1, 55, 343, 100};
        quickSort(array, 0, array.length - 1);
        System.out.println(Arrays.toString(array));
    }
}
